### PR TITLE
Fix GUI.do_trait_later to not sendPostedEvents

### LINKED ISSUE #272: On Qt, GUI.start_event_loop processes posted events before the event loop starts.
The [implementation](https://github.com/enthought/pyface/blob/5.1.0/pyface/ui/qt4/gui.py#L104-L115) of `GUI.start_event_loop` on Qt looks like this:
```
    def start_event_loop(self):
        if self._splash_screen is not None:
            self._splash_screen.close()

        # Make sure that we only set the 'started' trait after the main loop
        # has really started.
        self.set_trait_later(self, "started", True)

        logger.debug("---------- starting GUI event loop ----------")
        start_event_loop_qt4()

        self.started = False
```
But with the current implementation, `set_trait_later` can trigger the trait change immediately, thanks to a [`sendPostedEvents`](https://github.com/enthought/pyface/blob/5.1.0/pyface/ui/qt4/gui.py#L172-L173) call, so that code reacting to the `started` trait cannot assume that the event loop is running.

One particular situation in which this is an issue is when the code that reacts to `started` decides to stop the event loop (for example, because of some missing resource): we then end up with a situation where `gui.stop_event_loop` is called (a) outside the event loop, and (b) before `gui.start_event_loop` is called.

### PR DESCRIPTION
**Pull request by mdickinson**
Fixes #272.

### CODE REVIEW

File: pyface/ui/qt4/tests/test_gui.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+        def exit_app():
+            # Record whether the event loop is running or not, then exit.
+            application_running.append(
+                is_event_loop_running_qt4()
+            )
+            application.stop()
+
+        application.on_trait_change(exit_app, 'application_running')
+        application.start()
```
Comment (mdickinson) [inline pyface/ui/qt4/tests/test_gui.py]: Unpleasant aspect of this test: if anything goes wrong, it'll likely hang forever. That's already what happens on Windows if this test is run without the `gui.py` change.

**Comment by mdickinson** [comment]
I'll see if I can rework the test to not hang forever if something goes wrong.

**Comment by mdickinson** [comment]
> I'll see if I can rework the test to not hang forever if something goes wrong.

I don't think I can: the cure is worse than the disease.

**Comment by mdickinson** [comment]
> I'll see if I can rework the test to not hang forever if something goes wrong.

Done, and tested on Windows.

**Comment by corranwebster** [comment]
Basic idea looks good.

Is there any clean-up that needs to be done in the case of failure, such as flushing the event queue, so that we don't get interference with other tests?

**Comment by mdickinson** [comment]
I'll add a flush; there's probably not a whole lot more we can do. The main priority would be making sure that we're not introducing test interactions in the passing case.

**Comment by corranwebster** [comment]
LGTM. Merging.

**Pull request merged**